tracee/event_monitor_ebpf.c

```diff
@@ -1360,188 +1379,222 @@ int trace_ret_##name(struct pt_regs *ctx)                               \
+        if (!event_chosen(id) && !event_chosen(RAW_SYS_ENTER))
+            return 0;
+    }
+
+    if (event_chosen(RAW_SYS_ENTER)) {
+        context_t context = {};
+        init_context(&context);
+        set_buf_off(SUBMIT_BUF_IDX, sizeof(context_t));
+        buf_t *submit_p = get_buf(SUBMIT_BUF_IDX);
+        if (submit_p == NULL)
+            return 0;
+
+        context.eventid = RAW_SYS_ENTER;
+        context.argnum = 1;
+        context.retval = 0;
+
+        save_context_to_buf(submit_p, (void*)&context);
+
+        save_to_submit_buf(submit_p, (void*)&id, sizeof(int), INT_T, TAG_SYSCALL);
+        events_perf_submit(ctx);
+    }
+
+    // call syscall handler, if exists
+    sys_enter_tails.call(ctx, id);
+    return 0;
+}
+
+// include/trace/events/syscalls.h:
+// TP_PROTO(struct pt_regs *regs, long ret)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+TRACEPOINT_PROBE(raw_syscalls, sys_exit)
+#else
+RAW_TRACEPOINT_PROBE(sys_exit)
+#endif
+{
+    int id;
+    long ret;
+
+    // we don't support ia32 syscalls at this point
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    if (task->thread_info.status & TS_COMPAT)
         return 0;
 
-    context.eventid = RAW_SYSCALLS;
-    context.argnum = 1;
-    context.retval = 0;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+    void *ctx = args;
+    id = args->id;
+    ret = args->ret;
+#else
+    struct pt_regs *regs = (struct pt_regs*)ctx->args[0];
+    id = regs->orig_ax;
+    ret = ctx->args[1];
+#endif
 
-    save_context_to_buf(submit_p, (void*)&context);
+    // essential events should always be traced
+    if (id != SYS_CLONE && id != SYS_FORK && id != SYS_VFORK && id != SYS_EXECVE && id != SYS_EXECVEAT) {
+        if (!should_trace())
+            return 0;
+
+        if (!event_chosen(id) && !event_chosen(RAW_SYS_EXIT))
+            return 0;
+    }
+
+    if (event_chosen(RAW_SYS_EXIT)) {
+        context_t context = {};
+        init_context(&context);
+        set_buf_off(SUBMIT_BUF_IDX, sizeof(context_t));
+        buf_t *submit_p = get_buf(SUBMIT_BUF_IDX);
+        if (submit_p == NULL)
+            return 0;
+
+        context.eventid = RAW_SYS_EXIT;
+        context.argnum = 1;
+        context.retval = ret;
+
+        save_context_to_buf(submit_p, (void*)&context);
+
+        save_to_submit_buf(submit_p, (void*)&id, sizeof(int), INT_T, TAG_SYSCALL);
+        events_perf_submit(ctx);
+    }
 
-    save_to_submit_buf(submit_p, (void*)&(args->id), sizeof(int), INT_T, TAG_SYSCALL);
-    events_perf_submit((struct pt_regs *)args);
-    
+    // call syscall handler, if exists
+    save_retval(ret, id);
+    sys_exit_tails.call(ctx, id);
+    del_retval(id);
+    del_args(id);
     return 0;
 }
 
-int syscall__execve(struct pt_regs *ctx,
-    const char __user *filename,
-    const char __user *const __user *__argv,
-    const char __user *const __user *__envp)
+int syscall__execve(void *ctx)
```

Comment:
can this (and following kprobes) be moved into the raw/tracepoint handler? (if id==execve)

Reply:
The only syscalls kprobes that are left here are execve+execveat.
I'm indeed changing this in my next PR.
Pay attention, though, that instruction count limit is 4096, so we will still have to use tail calls for these two.

Reply:
this limit is per probe?

Reply:
Yes